sys/include/net/sock/async/event.h

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (C) 2019 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup        net_sock_async_event    Asynchronous sock with event API
+ * @ingroup         net_sock
+ * @brief           Provides an implementation of asynchronous sock for
+ *                  @ref sys_event
+ * @experimental    This API is still under development and should not be used
+ *                  in production yet.
+ *
+ * How To Use
+ * ----------
+ *
+ * You need to [include](@ref including-modules) at least one module that
+ * implements a [`sock` API](@ref net_sock) and the module `sock_async_event` in
+ * your application's Makefile.
+ *
+ * For the following example [`sock_udp`](@ref net_sock_udp) is used. It is
+ * however easily adaptable for other `sock` types:
+ *
+ * ### An asynchronous UDP Echo server using the event API
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * #include <stdio.h>
+ *
+ * #include "net/sock/udp.h"
+ * #include "net/sock/async/event.h"
+ *
+ * event_queue_t queue;
+ * uint8_t buf[128];
+ *
+ * void handler(sock_udp_t *sock, sock_async_flags_t type)
+ * {
+ *     if (type & SOCK_ASYNC_MSG_RECV) {
+ *         sock_udp_ep_t remote;
+ *         ssize_t res;
+ *
+ *         if ((res = sock_udp_recv(sock, buf, sizeof(buf), 0,
+ *                                  &remote)) >= 0) {
+ *             puts("Received a message");
+ *             if (sock_udp_send(sock, buf, res, &remote) < 0) {
+ *                 puts("Error sending reply");
+ *             }
+ *         }
+ *     }
+ * }
+ *
+ * int main(void)
+ * {
+ *     sock_udp_ep_t local = SOCK_IPV6_EP_ANY;
+ *     sock_udp_t sock;
+ *
+ *     local.port = 12345;
+ *
+ *     if (sock_udp_create(&sock, &local, NULL, 0) < 0) {
+ *         puts("Error creating UDP sock");
+ *         return 1;
+ *     }
+ *
+ *     event_queue_init(&queue);
+ *     sock_udp_event_init(&sock, &queue, handler);
+ *     event_loop(&queue);
+ *     return 0;
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * Above you see a simple UDP echo server using @ref sys_event. Don't forget to
+ * also @ref including-modules "include" the IPv6 module of your networking
+ * implementation (e.g. `gnrc_ipv6_default` for @ref net_gnrc GNRC) and at least
+ * one network device.
+ *
+ * After including the header file for @ref net_sock_udp "UDP sock" and
+ * @ref net_sock_async_event "asynchronous handling", we create the event queue
+ * `queue` and allocate some buffer space `buf` to store the data received by
+ * the server:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * #include "net/sock/udp.h"
```

Comment:
This doesn't seem to be needed...
```suggestion
```

Reply:
Yes and no. It is not really needed, as `net/sock/async.h` also pulls it in. However, for future proofness and just good style (functions and types from that header are used here after all) I would prefer to keep it in.

Reply:
Ok, I don't mind :)